### PR TITLE
Allow defined clients to use all endpoints

### PR DESCRIPTION
**Pull request by lukidzi**

### CODE REVIEW

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotProperties.kt [hunk 1]
```diff
@@ -70,6 +70,7 @@ class IncomingPermissionsProperties {
     var sourceIpAuthentication = SourceIpAuthenticationProperties()
     var selectorMatching: MutableMap<Client, SelectorMatching> = mutableMapOf()
     var tlsAuthentication = TlsAuthenticationProperties()
+    var allowedClientsForAllEndpoints = mutableListOf<String>()
```
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotProperties.kt:73]: Shouldn't it be `allowed to`?

Comment (pzmi) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotProperties.kt:73]: Also `clients allowed to` sound correct to me rather than `allowed clients to`

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt [hunk 2]
```diff
@@ -39,6 +39,9 @@ class RBACFilterFactory(
 
     private val anyPrincipal = Principal.newBuilder().setAny(true).build()
     private val denyForAllPrincipal = Principal.newBuilder().setNotId(anyPrincipal).build()
+    private val allowAllEndpointsForClients = incomingPermissionsProperties.allowedClientsForAllEndpoints.map {
```
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt:42]: Isn't it a list of clients? `clientsAllowedToAllEndpoints` or maybe something less mouthful `clientsWithFullAccess` or shorter `fullAccessClients` (alternatively replace `full` with `open` but I think `full` has more emphasis)

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt [hunk 3]
```diff
@@ -195,7 +198,7 @@ class RBACFilterFactory(
             roles.find { it.name == clientOrRole.name }?.clients ?: setOf(clientOrRole)
         }
         // sorted order ensures that we do not duplicate rules
-        return clients.toSortedSet()
+        return (clients + allowAllEndpointsForClients).toSortedSet()
```
Comment (pzmi) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt:201]: I'm starting to get lost in this factory, there are so many conditions and corner cases. I think it's time for a refactor and extracting some patterns from this code.

**Review by pzmi — APPROVED** [review]

**Review by andrzejwaw — DISMISSED** [review]

File: envoy-control-core/src/main/resources/lua/ingress_rbac_logging.lua [hunk 4]
```diff
@@ -42,24 +51,45 @@ function first_header_value_from_list(header_list, handle)
...
+    return false
+end
+
 function envoy_on_response(handle)
     local rbacMetadata = handle:streamInfo():dynamicMetadata():get("envoy.filters.http.rbac")
+    local lua_metadata = handle:streamInfo():dynamicMetadata():get("envoy.filters.http.lua") or {}
+    local allowed_client = lua_metadata["request.info.allowed_client"] or false
 
-    if rbacMetadata == nil or rbacMetadata["shadow_engine_result"] ~= "denied" then
+    if rbacMetadata == nil then
```
Comment (slonka) [inline envoy-control-core/src/main/resources/lua/ingress_rbac_logging.lua:69]: Why can't we do:

```
if rbacMetadata == nil or rbacMetadata["shadow_engine_result"] ~= "denied" or allowed_client == false then
```

?

File: envoy-control-runner/src/main/resources/application.yaml [hunk 5]
```diff
@@ -17,4 +17,9 @@ management:
...
     prometheus.enabled: true
   endpoints.web.exposure.include: "*"
-  metrics.export.prometheus.enabled: true
\ No newline at end of file
+  metrics.export.prometheus.enabled: true
+
+spring:
+  jackson:
+    deserialization:
+      fail-on-unknown-properties: false
```
Comment (slonka) [inline envoy-control-runner/src/main/resources/application.yaml:25]: Why is this needed?

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/permissions/IncomingPermissionsLoggingModeTest.kt [hunk 6]
```diff
@@ -691,20 +692,20 @@ internal class IncomingPermissionsLoggingModeTest : EnvoyControlTestConfiguratio
         val echo2Response = callEnvoyIngress(
             envoy = echo2Envoy,
             path = "/log-unlisted-clients",
-            headers = mapOf("x-service-name" to "service-name-from-header"),
-            useSsl = true,
+            headers = mapOf("x-service-name" to "allowed-client"),
```
Comment (slonka) [inline envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/permissions/IncomingPermissionsLoggingModeTest.kt:695]: Why are you changing this test and not adding another one?

**Review by slonka — CHANGES_REQUESTED** [review]
Some minor things and a couple of questions.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt [hunk 7]
```diff
@@ -254,7 +257,7 @@ class RBACFilterFactory(
             val principals = it.value.map { ipWithPrefix ->
                 val (ip, prefixLength) = ipWithPrefix.split("/")
 
-                Principal.newBuilder().setSourceIp(CidrRange.newBuilder()
+                Principal.newBuilder().setDirectRemoteIp(CidrRange.newBuilder()
```
[the remark this answers is not in the document]
Comment (lukidzi) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RBACFilterFactory.kt:260]: Yes, during tests I've found it is deprecated so I fixed that.

File: envoy-control-core/src/main/resources/lua/ingress_rbac_logging.lua (hunk 4 above)
Comment (lukidzi) [reply envoy-control-core/src/main/resources/lua/ingress_rbac_logging.lua:69]: Because, in this case if `rbacMetadata["shadow_engine_result"] ~= "denied"` then we would just return without any log. We want to log if `rbacMetadata["shadow_engine_result"] ~= "denied" and allowed_client == true` but in one mentioned by you there is `or` operator which need only one condition to pass.

File: envoy-control-runner/src/main/resources/application.yaml (hunk 5 above)
Comment (lukidzi) [reply envoy-control-runner/src/main/resources/application.yaml:25]: Tested something and didn't notice. I ll remove it

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/permissions/IncomingPermissionsLoggingModeTest.kt (hunk 6 above)
Comment (lukidzi) [reply envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/permissions/IncomingPermissionsLoggingModeTest.kt:695]: True, I've changed that while developing and forgot about it. I will add another test .a

File: envoy-control-core/src/main/resources/lua/ingress_rbac_logging.lua (hunk 4 above)
Comment (slonka) [reply envoy-control-core/src/main/resources/lua/ingress_rbac_logging.lua:65]: I simplified this for you: https://github.com/allegro/envoy-control/compare/allow_client...allow_client_slonka?expand=1

```lua
function envoy_on_response(handle)
    local rbacMetadata = handle:streamInfo():dynamicMetadata():get("envoy.filters.http.rbac") or {}
    local lua_metadata = handle:streamInfo():dynamicMetadata():get("envoy.filters.http.lua") or {}
    local allowed_client = lua_metadata["request.info.allowed_client"] or false

    if should_log_request(rbacMetadata, allowed_client) then
        log_request(lua_metadata, handle)
    end
end

function should_log_request(rbacMetadata, allowed_client)
    local is_shadow_denied = (rbacMetadata["shadow_engine_result"] or "") == "denied"

    if (not is_shadow_denied) and allowed_client == true then
        return true
    else
        return is_shadow_denied
    end
end
```

Comment (lukidzi) [reply envoy-control-core/src/main/resources/lua/ingress_rbac_logging.lua]: Thanks. Applied change :)

**Review by slonka — APPROVED** [review]

File: envoy-control-core/src/main/resources/lua/ingress_rbac_logging.lua [hunk 8]
```diff
@@ -42,24 +47,48 @@ function first_header_value_from_list(header_list, handle)
     return ""
 end
 
+function is_allowed_client(client_name, clients_allowed_to_all_endpoints)
+    for _,h in ipairs(clients_allowed_to_all_endpoints) do
+        if client_name ~= "" and h ~= "" and client_name == h then
```
Comment (pzmi) [inline envoy-control-core/src/main/resources/lua/ingress_rbac_logging.lua:52]: `client_name` condition could be moved outside of the loop.

**Review by pzmi — DISMISSED** [review]
Just one little comment

**Review by slonka — APPROVED** [review]

**Review by pzmi — APPROVED** [review]
